nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp

```diff
@@ -535,6 +565,46 @@ nav_msgs::msg::Path RegulatedPurePursuitController::transformGlobalPlan(
   return transformed_plan;
 }
 
+inline double RegulatedPurePursuitController::dot(double u_x, double u_y, double v_x, double v_y)
+{
+  // Supports only 2D
+  return  (u_x * v_x) + (u_y * v_y);
+}
+
+double RegulatedPurePursuitController::detectPathOrientation(
+  const geometry_msgs::msg::PoseStamped & pose )
+{
+  double path_direction_changed = 0.0;
+  double cusp_distance = 0.0;
+
+  // Iterating through the global path to determine the position of the cusp
+  for (auto iter = 1; iter <= (int)global_plan_.poses.size(); ++iter) {
+    // We have two vectors for the dot product OA and AB. Determining the vectors.
+    double oa_x = global_plan_.poses[iter].pose.position.x -
+      global_plan_.poses[iter - 1].pose.position.x;
+    double oa_y = global_plan_.poses[iter].pose.position.y -
+      global_plan_.poses[iter - 1].pose.position.y;
+    double ab_x = global_plan_.poses[iter + 1].pose.position.x -
+      global_plan_.poses[iter].pose.position.x;
+    double ab_y = global_plan_.poses[iter + 1].pose.position.y -
+      global_plan_.poses[iter].pose.position.y;
+
+    // Checking for the existance of cusp using the dot product
+    path_direction_changed = dot(oa_x, oa_y, ab_x, ab_y) > 0.0 ? 1.0 : -1.0;
+
+    // Getting the distance of the cusp, if the cusp is found
+    if (path_direction_changed == -1.0) {
```

Comment:
We don't actually use `path_direction_changed` for anything other than if detecting if negative. So you can remove that variable entirely and move its logic into the `if()` statement. Also you don't need to look at `-1.0` it just needs to be `< 0.0` so this could be written far more compactly as 

`if (dot(...) < 0.0)`

Reply:
Make sure that this is `<` not `<=`. If its 0.0, its not a reverse but a 90 degree move which isn't really a cusp worth doing this in.

Reply:
This is kind of the edge case, consider we have a straight path driving in reverse direction. Then the `findDirectionChange` function will return 0.0.

Then this [condition at this line](https://github.com/padhupradheep/navigation2/blob/e1ead123df1e85623c15e8141a26a06afb657109/nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp#L250) gets affected, which means that the lookahead point does not move at all.